src/common/transformations/src/transformations/mlir/mlir_op.cpp

```diff
@@ -267,6 +290,138 @@ namespace mlir {
+        case MLIR_MODE_DEFAULT:
+            return std::make_shared<MLIREvaluate>(std::move(module), mode);
+        default:
+            OPENVINO_THROW("Unsupported MLIR mode");
+    }
+}
+
+#ifdef GC_USE_IMEX // GC_GPU requires IMEX support
+
+cl_device_id extract_device_from_context(cl_context context) {
+    size_t devices_size;
+    cl_int err = clGetContextInfo(context, CL_CONTEXT_DEVICES, 0, NULL, &devices_size);
+    if (err != CL_SUCCESS) {
+        OPENVINO_THROW("Error getting context info: ", err);
+    }
+    if (devices_size / sizeof(cl_device_id) != 1) {
+        OPENVINO_THROW("Expected exactly one device in the context, got ", devices_size);
+    }
+
+    cl_device_id devices;
+    err = clGetContextInfo(context, CL_CONTEXT_DEVICES, devices_size, &devices, NULL);
+    if (err != CL_SUCCESS) {
+        OPENVINO_THROW("Error getting device IDs: ", err);
+    }
+
+    return devices;
+}
+
+
+
+MLIREvaluateGcGPU::MLIREvaluateGcGPU(OwningOpRef<mlir::ModuleOp> _module, std::shared_ptr<ov::EvaluationContext> loweringContext) {
+    OPENVINO_MLIR_DEBUG_PRINT(
+        "[ DEBUG ] Source MLIR:\n"
+        "-----------------------------------------\n");
+    OPENVINO_MLIR_DEBUG(_module->dump());
+    OPENVINO_MLIR_DEBUG_PRINT(
+        "-----------------------------------------\n");
+
+    gc::gpu::OclModuleBuilder builder(std::move(_module));
+
+    auto it = loweringContext->find(ov::intel_gpu::ocl_context.name());
+    if (it == loweringContext->end()) {
+        OPENVINO_THROW("No cl_context provided for OpenCL execution");
+    }
+    // QUESTION: are we sure that there's always only one device per context?
+    // QUESTION: may we pass context as void* directly to 'builder.build()' and make it responsible
+    // for extracting the device?
+    auto context = reinterpret_cast<cl_context>(it->second.as<ov::intel_gpu::gpu_handle_param>());
+    module = gcGetOrReport(builder.build(extract_device_from_context(context), context));
+
+    // gc::gpu::OclModule::dump is not implemented yet
+    // OPENVINO_MLIR_DEBUG_PRINT(
+    //     "[ DEBUG ] Target LLVM:\n"
+    //     "-----------------------------------------\n");
+    // OPENVINO_MLIR_DEBUG(module->dump());
+    // OPENVINO_MLIR_DEBUG_PRINT(
+    //     "-----------------------------------------\n");
+};
+
+bool MLIREvaluateGcGPU::invoke(std::vector<void*>& args, const ov::EvaluationContext& evaluationContext) {
+    gc::gpu::OclContext ctx = build_ocl_context(evaluationContext);
+    gc::gpu::StaticExecutor exec(module);
+
+    auto it = evaluationContext.find(ov::intel_gpu::memory_type::is_kernel_arg_usm.name());
+    if (it == evaluationContext.end()) {
+        OPENVINO_THROW("No is_kernel_arg_usm provided for OpenCL execution");
+    }
+    std::vector<bool> arg_types = it->second.as<std::vector<bool>>();
+
+    for (size_t i = 0; i < args.size(); ++i) {
+        exec.arg(args[i], arg_types[i]);
+    }
+    exec(ctx);
+    // Should we 'wait()' here?
+    // ctx.finish();
```

Comment:
I think we should return an event. @vladimir-paramuzov, do we need to return it somehow or can the plugin inspect the queue?

Reply:
We save events only for out of order queues.

Reply:
As far as I understand we can't construct OV's ocl_user_event from a `cl_event`, so returning an event generated by our runtime won't work.

OV creates an event on their side in [`intel_gpu/plugin/ops/mlir_op.cpp`](https://github.com/dchigarev/openvino/blob/dc78071ce15f9c481088a2bd42d4e49a89181d3e/src/plugins/intel_gpu/src/plugin/ops/mlir_op.cpp#L34), if GC-GPU runtime would support taking an already generated cl_event, then we could forward it via `EvaluationContext`.

@AndreyPavlenko is anything blocking us from accepting a `cl_event` in `OclContext` or in `exec`?

Reply:
The only problem that remains is [the one mentioned by](https://github.com/slyalin/openvino/pull/169#discussion_r1782954412) @AndreyPavlenko [above](https://github.com/slyalin/openvino/pull/169#discussion_r1782954412): the gpu runtime of graph compiler doesn't produce events in case of a sequential `cl_command_queue` at the moment (since it assumes that the whole queue will be waited and not individual events). 

[The logic at the beginning of `generic_primitive::execute_function`](https://github.com/slyalin/openvino/blob/21ed9ac767688e6038eafa6d2f842eea966920c8/src/plugins/intel_gpu/src/plugin/ops/mlir_op.cpp#L28-L35), where we only use events in case of an out-of-order queue, makes me wonder if we have to return a valid event to openvino when using a sequential queue.

What are your thoughts on that matter @vladimir-paramuzov? Are `cldnn:events` used anyhow in case of a sequential queue in openvino (if yes, we have to modify our gc-gpu runtime to produce `cl_event`s with any queue type), or we can safely ignore them assuming that the whole queue will be waited and not individual events?

Reply:
As I mentioned in the chat, events are needed only for OOOQ. in case of in order queue you can just return an object created by `stream.create_user_event(true);` call